### PR TITLE
[CS2113-T13-2] Domnus

### PR DESCRIPTION
**Pull request by johanesrafael**
Domnus is a portable timetable scheduler for NUS students and provides 3 main feature namely: Task Tracking, Timetabling and Cap Calculating features. This application aims to increase the productivity of students in the context of having an effective task tracking software product.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,407 @@
 # Developer Guide
 
-## Design & implementation
+## 1. Table of content
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:3]: Consider using links to the different sections for the table of contents? It will ensure easier navigability of the document.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,407 @@
...
 
+1.  Ensure that you have Java 11 or above installed.
+2.  Download the latest version of  `Duke`  from  [Our Release Page](https://github.com/AY2021S1-CS2113-T13-2/tp/releases/tag/v1.0).
+3.  Copy the file to the folder you want to use as the home folder for your Mobile Nusmod.
+4.  Open the Command Prompt if you are running on Windows or Terminal if you are running on Mac OS.
+5.  Navigate to your home folder and type ‘java -jar domnus.jar’
+6.  Type ‘bye’ to terminate your session.
 
-## Product scope
+## 3. Design
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md]: Consider adding 1-2 lines preface for this section, to introduce the points covered here.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,407 @@
...
+
+### 3.1 Architecture
+The **Architecture Diagram** below represents a high-level design overview of the App. Specifically, it is done with an **N-tier architectural style**, where the higher layers make use of services provided by lower layers. 
+
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:63]: Nicely described system architecture!

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,407 @@
...
+
+**3.2 Main Layer**<br>
+For the `main` layer, it contains a single class known as `Duke`. 
+
+**3.3 UI Layer**<br>
+Main gets user input and displays messages through the use of UI component. 
+The UI layer entails the package *visualize*, which contains classes *ColoredString*, *Bitmap*, *UI*, *Cli*, 
+*FancyCli* and enumerations *Color* and *Sprite* in the following structure:
+![uml](Images/Package%20visualize.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:79]: Perhaps use the standard notation for class diagrams? Instead of icons for class, interface, enum etc. it would be good if it follows the textbook notation. Eg. Instead of the icon E in shown for the enumerations, using <<enumeration>> instead..

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,407 @@
...
+
+**1.4 Execute Layer**<br>
+Once CI processed the user input, duke proceeds to redirect the input to Execute for execution of action. 
+
+**1.5 Storage Layer**<br>
+Once CI processed the user input, duke proceeds to redirect the input to Execute for execution of action. 
+
+**1.6 Flow of DOMSUN**<br>
+The sequence diagram below shows the main interaction of classes with each other throughout the whole lifecycle of DOMSUM.
+![uml](Images/DOMSUM_Main_Flow.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:114]: Some thing about the sequence diagram:
* The activation bars begin before the call arrows and end after the return arrows, which is an incorrect notation. Consider changing the activation bars to begin at the arrow call and end at the return call?
* The activation bar for `:DOMNUS` in this diagram ends within the loop itself, thus implying the program execution terminates within the loop. I'd suggest extending the activation bar a bit more?

Otherwise a nicely done diagram!

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,407 @@
...
+|-----|---|---|------|
+|blah|blah|blah|blah|
+
+A typical flow of execution would be:
+1. blah
+1. blah
+1. blah
+
+### 4.2 Checker Feature 
+![here](Images/Checker_Diagram.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md]: I think some the class names are missing the ":" in front of them. Not sure if this is an optional/necessary convention but perhaps consider adding the colons?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 2. `CalculateCapAction#act()` retrieves data from the stored user's grades.
+
+Step 3. `CalculateCapAction#act()` then retrieves module data from the `modulelist.txt` to determine Modular Credit (MC) allocation.
+
+Step 4. CAP value is calculated and returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![cap uml diagram](Images/CalculateCapSequence.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md]: I think its a glitch of some sort, but the background of the image seems to be transparent so it is appearing black so I can't seem much in the diagram. 
image

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 2. `ReminderAction#act` retrieves tasklist data from the user's list
+
+Step 3. `ReminderAction#act` then sorts the due dates in ascending order
+
+Step 4: Tasks due within 3 days are returned to the user through Ui
+
+The following sequence diagram diagram shows how the reminder operation works
+
+![Reminder_Sequence_Diagram](Images/ReminderAction_Sequence_Diagram.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:274]: Same issue as the previous diagram :(

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 1. The user executes `remind [time]` command to set schedule for the reminder. Command is then parsed by `RemindAction#prepare()` to be passed as arguments for `RemindAction#act()`.
+
+Step 2. `RemindAction#act()` calls `RemindAction#getSchedule` to pass the schedule later in `Ui`.
+
+Step 3. The schedule is returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Postpone_Sequence_Diagram](Images/Remind.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md]: The first call arrow in this sequence diagram starts at the middle of the activation bar, and there is no return arrow from the function call to represent transfer of control back to the caller.

image

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,407 @@
...
 ### Value proposition
+All in one app to track tasks and their dates, monitor productivity and calculate cap. 
 
-{Describe the value proposition: what problem does it solve?}
+## Appendix B. User Stories
 
-## User Stories
-
-|Version| As a ... | I want to ... | So that I can ...|
+|Priority| As a ... | I want to ... | So that I can ...|
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:374]: A nice and comprehensive list of user stories!

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,407 @@
...
+_{More to be added}_
+
+## Appendix E. Glossary
+
+ - N-tier Architectural Style 
+	 - In the n-tier style, high layers make use of services provided by lower layers. Lower layers are independent of higher layers. 
+ - Mainstream OS: Windows, Linux, Unix, OS-X
+ - Private contact detail 
+
+## Appendix F. Instructions for manual testing
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:438]: This part is well written! :D

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 2. `PostponeAction#act()` re-sets the date of the targeted task from the stored user's tasks by default a day.
+
+Step 3. `PostponeAction#act()` then updates the stored user's data.
+
+Step 4. Postponed target task is returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Postpone_Sequence_Diagram](Images/PostponeAction_Sequence_Diagram.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:316]: A fairly well done sequence diagram! However, I didn't quite get what was happening in this part:
image
The control is returned not the calling class, but the other class? Consider doing it like this maybe?: 
image

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 2. `GradeAction#act()` takes in data prepared by `GradeAction#prepare()`.
+
+Step 3. `GradeAction#act()` then retrieves module data from the `modulelist.txt` to determine module details.
+
+Step 4. Grade is attributed to the corresponding modules and response message is returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Grade_Sequence_Diagram](Images/GradeSequence.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md]: Similar transparent background issue, perhaps it might've occurred whilst rendering the image?

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,407 @@
...
+
+Step 1. The user executes `focus [task type]` command to filter based on task type. Command is then parsed by `FocusAction#prepare()` to be passed as arguments for `FocusAction#act()`.
+
+Step 2. `FocusAction#act()` then sets flag in the data.
+
+Step 3. Changed context is returned to inform the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Postpone_Sequence_Diagram](Images/Focus.png)
```
Comment (sevenseasofbri) [inline docs/DeveloperGuide.md:358]: I think a similar issue with the activation bars starting before the arrow here?

**Review by sevenseasofbri — COMMENTED** [review]
### Overall, a very well done Dev Guide! 💯 
Just a few suggestions:
* Links to the ToC leading to each section.
* Figure numbers.
One thing I noticed was that the sequence diagrams were probably done using different softwares as they're all sort of different from each other. Perhaps standardize the diagrams?

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,513 @@
...
+**7. Appendix C Use Cases** \
+**8. Appendix D Non-funcitonal Requirements** \
+**9. Appendix E Glossary** \
+**10. Appendix F. Instruction for Manual Testing**
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+## 2. Setting Up
 
+1.  Ensure that you have Java 11 or above installed.
+2.  Download the latest version of  `Duke`  from  [Our Release Page](https://github.com/AY2021S1-CS2113-T13-2/tp/releases/tag/v1.0).
```
Comment (Lusi711) [inline docs/DeveloperGuide.md]: Do remember to update the link to the release page

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,513 @@
...
+
+**1.4 Execute Layer**<br>
+Once CI processed the user input, duke proceeds to redirect the input to Execute for execution of action. 
+
+**1.5 Storage Layer**<br>
+Once CI processed the user input, duke proceeds to redirect the input to Execute for execution of action. 
+
+**1.6 Flow of DOMSUN**<br>
+The sequence diagram below shows the main interaction of classes with each other throughout the whole lifecycle of DOMSUM.
+![uml](Images/DOMSUM_Main_Flow.png)
```
Comment (Lusi711) [inline docs/DeveloperGuide.md:114]: The layout of the image looks wrong on the website. Could it be solved with a whitespace between the text and the image?

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,34 +1,513 @@
...
+
+## 4. Implementation<br>
+This section highlights some of our project's key feature and its implementation. 
+
+### 4.1 Take Feature
```
Comment (Lusi711) [inline docs/DeveloperGuide.md]: Would it be more easy to explain with a sequence diagram?

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,513 @@
...
+
+Step 2. Data instantiates a Checker with its existing list by calling its constructor, and the task to be added to the list 
+
+Step 3. Data proceeds to call the `checkRecurrenceDate(Task)` of the Checker class, to get a newDate if today's date is beyond the stated weekly date.
+ 
+Step 4. If `newDate` is not `null`it shows that there is a new updated date. Therefore, we proceed to update the object with our updated weekly date. 
+
+Step 5. Now we proceed to call `checkDuplicates()` of Checker class. 
+
+Step 6. If `false` , there is no duplicates in the existing list, and the task can be safely added. Otherwise, no action will be taken.
```
Comment (Lusi711) [inline docs/DeveloperGuide.md]: Very clear explanation!

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,513 @@
...
+
+Step 1. The user executes `snooze` command to snooze for the reminder. 
+
+Step 2. `SnoozeAction#getNewInterval` sets and returns the new interval.
+
+Step 3. The new interval is returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Postpone_Sequence_Diagram](Images/Snooze.png)
```
Comment (Lusi711) [inline docs/DeveloperGuide.md:294]: Activation bar of method getNewInterval() starts before the method call arrives in this diagram

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -132,14 +132,13 @@ Step 4. CAP value is calculated and returned to the user through `Ui`.
 
 The following activity diagram summarizes what happens when a user executes a new command:
 
-![cap uml diagram](Images/Cap_Calculator_Diagram.JPG)
-
+![cap uml diagram](Images/CalculateCapSequence.png)
```
Comment (WYing333) [inline docs/DeveloperGuide.md]: Cannot see the lines and descriptions of the diagrams clearly. Maybe can change the color of the background or the lines.
diagram

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -177,7 +176,28 @@ The following activity diagram summarizes what happens when a user executes a ne
...
+
+Step 2. `GradeAction#act()` takes in data prepared by `GradeAction#prepare()`.
+
+Step 3. `GradeAction#act()` then retrieves module data from the `modulelist.txt` to determine module details.
+
+Step 4. Grade is attributed to the corresponding modules and response message is returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![Grade_Sequence_Diagram](Images/GradeSequence.png)
```
Comment (WYing333) [inline docs/DeveloperGuide.md]: Same issue as the one described above.
diagramm

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -214,7 +272,7 @@ The following activity diagram summarizes what happens when a user executes a ne
 
 ![Postpone_Sequence_Diagram](Images/PostponeAction_Sequence_Diagram.png)
```
Comment (WYing333) [inline docs/DeveloperGuide.md:316]: Overall, the diagrams are quite clear. However, maybe the way to draw diagrams could be consistent.

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -192,7 +250,7 @@ The following activity diagram summarizes what happens when a user executes a ne
 
 ![Postpone_Sequence_Diagram](Images/Snooze.png)
```
Comment (WYing333) [inline docs/DeveloperGuide.md:294]: The arrow of the method getNewInterval() needs to be at the start of the activation bar.
bar

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md:114]: This diagram have some indentation before it, which is not consistent with the other ones. And I found it's because there is a line "lifecycle of DOMSUM." before the diagram. So please try separate them.
image

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -1,34 +1,513 @@
...
+
+Step 2. `CalculateCapAction#act()` retrieves data from the stored user's grades.
+
+Step 3. `CalculateCapAction#act()` then retrieves module data from the `modulelist.txt` to determine Modular Credit (MC) allocation.
+
+Step 4. CAP value is calculated and returned to the user through `Ui`.
+
+The following activity diagram summarizes what happens when a user executes a new command:
+
+![cap uml diagram](Images/CalculateCapSequence.png)
```
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md]: This diagram, as well as the next one for 4.5 Reminder Feature, are all have a black(or transparent) background, which makes it quite hard to follow  the lines and arrows. Please check if you exported the UML in a proper way and try to solve it.
image

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (Johnson-Yee) [inline docs/DeveloperGuide.md:114]: Hey there, some comments about the sequence diagram here.
1. The call arrows should begin at the top of the activation bar. The return arrows should be at the bottom fo the activation bar.
2. I noticed that there is also an alt block for isFancy, not sure of its use there as there is no activation block in it or no method called.
3. After the return arrow of the savefile, is there a reason for the extension of activation bar for :DomNUS?

Comment (Johnson-Yee) [inline docs/DeveloperGuide.md:114]: In addition to the points made previously, the description of some arrows could be changed eg there are two arrows stating "Parser", not sure of its meaning, isit calling a function?
Otherwise, the diagram looks good!

File: docs/DeveloperGuide.md [hunk 25]
```diff
@@ -1,34 +1,513 @@
 # Developer Guide
 
-## Design & implementation
+## 1. Table of content
```
Comment (Johnson-Yee) [inline docs/DeveloperGuide.md:3]: Slight error here: Table of contents instead of content.
Do make sure the links work as well!

File: docs/DeveloperGuide.md [hunk 26]
```diff
@@ -1,34 +1,513 @@
...
+| ** |Student before start of semester|Find the modules either by keyword, module code or even MC |Easily see the desired modules |
+| ** |Student before start of semester|Select the modules but not taking it yet |Easily whether the MC fits my requirement |
+| *  |Student before start of semester|Find out the etails of the Module|To find out more about the modules.|
+| *** |Student before start of semester|Take the desired modules|Mark the modules that i want to take as taken  |
+| *** |Student during the semester|Add tasks such as todo,deadline and event into my list|Easily keep track of all the task i have to complete  |
+| ** |Student during the semester|Have a Reminder of which deadline is due soon|Ensure that no task is missed out  |
+| *** |Student during the semester|Add task to modules |Easily know which tasks belongs to which modules  |
+| *** |Student during the semester|Delete task once they are completed |Remove unnecessary task on the list |
+| ** |Student after the semester|Calculate the CAP of my individual modules|Easily find out my performance this semester |
+| * |Student after the semester|Clear the list of tasks and modules|Start afresh for the next semester |
```
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md]: I can see that currently you split the user stories into three groups which is good. But as you have already prioritized them, I suggest that you could reorder them by their priorities.
For example:
*** before
**  before
*** during
** during
...

File: docs/DeveloperGuide.md [hunk 27]
```diff
@@ -1,34 +1,513 @@
...
+
+The checker mechanism is facilitated by the utility class `Checker`. It is an independent class on its own without extensions and is stored under the `Data` package of our app. The class implements the following operations: 
+
+ - `checkDuplicates()`- Calls the checkClash method and return the status of boolean variable `isClash` .
+ - `checkClash(ArrayList< item >, Item)`- Updates `isClash` once a duplicate item is found in the list.
+ - `checkRecurrenceDate(Task)` - Checks if the current date is beyond the stated date in the list, and provides a new update for the date recurring date.
+
+Given below is an example usage scenario and how the checker mechanism behaves at each step. 
+
+![here](Images/Checker_Diagram.png)
```
Comment (Johnson-Yee) [inline docs/DeveloperGuide.md]: Could include activation bars in your sequence diagram.
Otherwise, good job!
image

File: docs/DeveloperGuide.md [hunk 28]
```diff
@@ -1,34 +1,513 @@
...
+*FancyCli* and enumerations *Color* and *Sprite* in the following structure:
+![uml](Images/Package%20visualize.png)
+
+UI's interaction with user<br>
+UI gets user input through *nextline()*, and renders strings as a user-comprehensible interface through 
+*update(String, Data)*.
+
+UI's interaction with the rest of the program<br>
+UI passes the user's input string out to the Duke object, which then passes the string to the Command Interpreter layer.
+UI also reads data from the Data object for refreshing purposes, but does not modify it.
```
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md]: Maybe it's better to rearrange this part. It looks a little bit messy.
And for these two *nextline()* and *update(String, Data)* do you mean `nextline()` and `update(String, Data)` instead?

File: docs/DeveloperGuide.md (hunk 24 above)
Comment (Johnson-Yee) [inline docs/DeveloperGuide.md]: Same problem encountered as the above reviewers causing the diagram to be messy. Maybe there was a problem with the picture file.

File: docs/DeveloperGuide.md [hunk 29]
```diff
@@ -1,34 +1,513 @@
...
+Step 3. Next, execute layer will call `action.act()` which causes StatsAction to begin its `act()` operation. If `isMod` flag is set, `act()` will search for the user specified module and get the list of tasks tagged to it.
+
+Step 5. Once the list of task is obtain, the operation will loop through the task list and count the number of completed task followed by generating a ratio. 
+
+Step 6. This ratio will be passed into `roundedRatioBar` to return *String* of a rounded ratio to 1 decimal place enclosing it in square brackets. 
+
+Step 7. Now `StatsAction` is completed and it will return this string back to `Execute` for to be printed through `UI`. 
+
+
+**Design consideration:**
```
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md:223]: You could delete this line if there is no content.

File: docs/DeveloperGuide.md [hunk 30]
```diff
@@ -1,34 +1,513 @@
...
+Step 3. Data proceeds to call the `checkRecurrenceDate(Task)` of the Checker class, to get a newDate if today's date is beyond the stated weekly date.
+ 
+Step 4. If `newDate` is not `null`it shows that there is a new updated date. Therefore, we proceed to update the object with our updated weekly date. 
+
+Step 5. Now we proceed to call `checkDuplicates()` of Checker class. 
+
+Step 6. If `false` , there is no duplicates in the existing list, and the task can be safely added. Otherwise, no action will be taken. 
+
+**Design consideration: 
+Aspect: How checker executes**
```
Comment (Wu-Haitao) [inline docs/DeveloperGuide.md]: This part looks problematic on the page. I guess you forgot to add "***" at the right places?
image

**Review by Wu-Haitao — COMMENTED** [review]
The DG is generally good. 👍
